### PR TITLE
server: reduce context switch about batch commands

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Reduce context switch about batch commands.
* let background threads notify gRPC threads directly instead of bypass an extra thread.
* remove the wait strategy in collecting responses to clients

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
make dev.

### CPU usage
the origin version:
图片

with the patch (the right one):
图片

### context-switch
the origin version:
图片

with the patch (the right one):
图片

### Conclusions
* Remove the thread can reduce CPU usage
* Remove the thread can reduce context switch

### CODE REVIEW

**Comment by siddontang** [comment]
can you paste the benchmark result here?

**Comment by hicqu** [comment]
PTAL @zhangjinpeng1987 @Connor1996

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -914,48 +900,62 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+        } else {
+            return;
+        }
+        *s = None;
+    }
+}
+
+fn poll_future_notify<F: Future<Item = (), Error = ()> + Send + 'static>(f: F) {
+    let spawn = Arc::new(Mutex::new(Some(executor::spawn(f))));
+    let notify = BatchCommandsNotify(Arc::clone(&spawn));
```
Comment (zhangjinpeng1987) [inline src/server/service/kv.rs]: Why clone spawn since it only used here?

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -914,48 +900,62 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
-    let executor1 = executor.clone();
     let f = resp.and_then(move |resp| {
-        if tx.send((id, resp)).is_err() {
+        if tx.send_and_notify((id, resp)).is_err() {
             error!("KvService response batch commands fail");
             return Err(());
         }
         timer.observe_duration();
-        if let Some(notifier) = tx.get_notifier() {
-            if thread_load.in_heavy_load() {
```
Comment (zhangjinpeng1987) [inline src/server/service/kv.rs]: in_heavy_load will not used any more?

Comment (hicqu) [reply src/server/service/kv.rs]: It's still used to tell clients TiKV's load.

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -904,48 +890,58 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+    }
+}
+impl<F> Notify for BatchCommandsNotify<F>
+where
+    F: Future<Item = (), Error = ()> + Send + 'static,
+{
+    fn notify(&self, id: usize) {
+        let n = Arc::new(self.clone());
+        let mut s = self.0.lock().unwrap();
+        match s.as_mut().map(|spawn| spawn.poll_future_notify(&n, id)) {
```
Comment (overvenus) [inline src/server/service/kv.rs]: Will `notify ` be called in the `poll_future_notify ` recursively?

File: components/tikv_util/src/mpsc/batch.rs [hunk 4]
```diff
@@ -120,6 +120,13 @@ impl<T> Sender<T> {
         Ok(())
     }
 
+    #[inline]
+    pub fn send_and_notify(&self, t: T) -> Result<(), SendError<T>> {
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/tikv_util/src/mpsc/batch.rs]: `send` won't notify pollers for every message, but `send_and_notify` does.

File: src/server/service/kv.rs [hunk 5]
```diff
@@ -904,48 +890,58 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+    fn clone(&self) -> BatchCommandsNotify<F> {
+        BatchCommandsNotify(Arc::clone(&self.0))
+    }
+}
+impl<F> Notify for BatchCommandsNotify<F>
+where
+    F: Future<Item = (), Error = ()> + Send + 'static,
+{
+    fn notify(&self, id: usize) {
+        let n = Arc::new(self.clone());
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/server/service/kv.rs]: No, it's required by `poll_futurue_noitfy` API.

File: src/server/service/kv.rs (hunk 3 above)
Comment (hicqu) [reply src/server/service/kv.rs]: `poll_future_notify` will call `notify` once, no recursion.

File: src/server/service/kv.rs [hunk 6]
```diff
@@ -842,29 +839,18 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
         sink: DuplexSink<BatchCommandsResponse>,
     ) {
         let (tx, rx) = unbounded(GRPC_MSG_NOTIFY_SIZE);
-        let executor = self.collect_pool.clone();
```
Comment (Connor1996) [inline src/server/service/kv.rs]: should remove `collect_poll` from `Service` too?

File: src/server/service/kv.rs [hunk 7]
```diff
@@ -1846,4 +1842,48 @@ mod tests {
...
+        let (tx, rx) = oneshot::channel::<usize>();
+        thread::Builder::new()
+            .name("source".to_owned())
+            .spawn(move || {
+                tx.send(100).unwrap();
+            })
+            .unwrap();
+
+        let (tx1, rx1) = oneshot::channel::<usize>();
+        thread::sleep(time::Duration::from_secs(3));
```
Comment (zhangjinpeng1987) [inline src/server/service/kv.rs]: Better to use channel sync.

**Pull request merged**